src/core/PlayerInfo.cpp

```diff
@@ -0,0 +1,662 @@
+	}
+
+	m_fRoadDensity = clamp(m_fRoadDensity, 0.4f, 1.45f);
+
+	// Because vehicle enter/exit use same key binding.
+	bool enterOrExitVeh;
+	if (m_pPed->bVehExitWillBeInstant && m_pPed->bInVehicle)
+		enterOrExitVeh = CPad::GetPad(0)->ExitVehicleJustDown();
+	else
+		enterOrExitVeh = CPad::GetPad(0)->GetExitVehicle();
+
+	if (enterOrExitVeh && m_pPed->m_nPedState != PED_SNIPER_MODE && m_pPed->m_nPedState != PED_ROCKET_MODE) {
+		if (m_pPed->bInVehicle) {
+			if (!m_pRemoteVehicle) {
+				CEntity *surfaceBelowVeh = m_pPed->m_pMyVehicle->m_pCurGroundEntity;
+				if (!surfaceBelowVeh || !CBridge::ThisIsABridgeObjectMovingUp(surfaceBelowVeh->GetModelIndex())) {
+					CVehicle *veh = m_pPed->m_pMyVehicle;
+					if (!veh->IsBoat() || veh->m_nDoorLock == CARLOCK_LOCKED_PLAYER_INSIDE) {
+
+						// This condition will always return true, else block was probably WIP Miami code.
+						if (veh->m_vehType != VEHICLE_TYPE_BIKE || veh->m_nDoorLock == CARLOCK_LOCKED_PLAYER_INSIDE) {
+							if (veh->GetStatus() != STATUS_WRECKED && veh->GetStatus() != STATUS_TRAIN_MOVING && veh->m_nDoorLock != CARLOCK_LOCKED_PLAYER_INSIDE) {
+								if (veh->m_vecMoveSpeed.Magnitude() < 0.17f && CTimer::GetTimeScale() >= 0.5f && !veh->bIsInWater) {
+									m_pPed->SetObjective(OBJECTIVE_LEAVE_CAR, veh);
+								}
+							}
+						} else {
+							CVector sth = 0.7f * veh->GetRight() + veh->GetPosition();
+							bool found = false;
+							float groundZ = CWorld::FindGroundZFor3DCoord(sth.x, sth.y, 2.0f + sth.z, &found);
+
+							if (found)
+								sth.z = 1.0f + groundZ;
+							m_pPed->SetPedState(PED_IDLE);
+							m_pPed->SetMoveState(PEDMOVE_STILL);
+							CPed::PedSetOutCarCB(0, m_pPed);
+							CAnimManager::BlendAnimation(m_pPed->GetClump(), m_pPed->m_animGroup, ANIM_STD_IDLE, 100.0f);
+							CAnimManager::BlendAnimation(m_pPed->GetClump(), ASSOCGRP_STD, ANIM_STD_FALL_LAND, 100.0f);
+							m_pPed->SetPosition(sth);
+							m_pPed->SetMoveState(PEDMOVE_STILL);
+							m_pPed->m_vecMoveSpeed = veh->m_vecMoveSpeed;
+						}
+					} else {
+						// The code in here was under CPed::SetExitBoat in VC, did the same for here.
+						m_pPed->SetExitBoat(veh);
+						m_pPed->bTryingToReachDryLand = true;
+					}
+				}
+			}
+		} else {
+			// Enter vehicle
+			if (CPad::GetPad(0)->ExitVehicleJustDown()) {
+				bool weAreOnBoat = false;
+				float lastCloseness = 0.0f;
+				CVehicle *carBelow = nil;
+				CEntity *surfaceBelow = m_pPed->m_pCurrentPhysSurface;
+				if (surfaceBelow && surfaceBelow->IsVehicle()) {
+					carBelow = (CVehicle*)surfaceBelow;
+					if (carBelow->IsBoat()) {
+						weAreOnBoat = true;
+						m_pPed->bOnBoat = true;
+#ifdef VC_PED_PORTS
+						if (carBelow->GetStatus() != STATUS_WRECKED && carBelow->GetUp().z > 0.3f)
+#else
+						if (carBelow->GetStatus() != STATUS_WRECKED)
+#endif
+							m_pPed->SetSeekBoatPosition(carBelow);
+					}
+				}
+				// Find closest car
+				if (!weAreOnBoat) {
```

Comment:
I'm thinking if `weAreOnBoat` is true, maybe Take-Two shouldn't frivolously take down community projects which fix and improve their terribly maintained games.

Just a thought.